airstack-modules/modules/airstack/social/social.ts

```diff
@@ -0,0 +1,262 @@
+      toAddress,
+      airUser.id,
+      airProfile.id,
+      profileName,
+      fromAddress,
+      toAddress,
+      tokenId,
+      tokenAddress,
+      airUserExtraIds,
+      airProfileExtraIds,
+    );
+    airUserRegisteredTransaction.save();
+  }
+
+  /**
+   * @dev this function does not create any entity, it just returns the air extra data entity ids
+   * @param extras air extra data array
+   * @returns air extra data entity ids
+   */
+  function getAirExtraDataEntityIds(
+    extras: AirExtra[],
+  ): string[] {
+    let entityIds = new Array<string>();
+    for (let i = 0; i < extras.length; i++) {
+      let extra = extras[i];
+      entityIds.push(extra.id);
+    }
+    return entityIds as string[];
+  }
+
+  /**
+   * @dev this function does not save the returned entity
+   * @dev this function gets or creates a AirUser entity
+   * @param chainId chain id
+   * @param block air block entity
+   * @param dappUserId dapp user id
+   * @param address air user address (owner of the dappUserId)
+  */
+  function getOrCreateAirUser(
+    chainId: string,
+    block: AirBlock,
+    dappUserId: string,
+    address: string,
+    extras: string[]
+  ): AirUser {
+    let id = chainId.concat('-').concat(dappUserId);
+    let entity = AirUser.load(id);
+    if (entity == null) {
+      entity = new AirUser(id);
+      let airAccount = getOrCreateAirAccount(chainId, address, block);
+      airAccount.save();
+      entity.dappUserId = dappUserId;
+      entity.address = airAccount.id;
+      entity.createdAt = block.id;
+      if (extras.length > 0) entity.extras = extras;
+      entity.lastUpdatedAt = block.id;
+    }
+    return entity as AirUser;
+  }
+
+  /**
+   * @dev this function does not save the returned entity
+   * @dev this function creates a AirProfile entity
+   * @param block ethereum block
+   * @param chainId chain id
+   * @param userId air user id
+   * @param name air profile name
+   * @param tokenId air profile token id
+   * @param tokenAddress air profile token address
+   * @param extraIds air extra data entity ids
+   * @returns air profile entity
+   */
+  function createAirProfile(
+    block: AirBlock,
+    chainId: string,
+    userId: string,
+    name: string,
+    tokenId: string,
+    tokenAddress: string,
+    extraIds: string[],
+  ): AirProfile {
+    let id = userId.concat("-").concat(tokenAddress).concat("-").concat(tokenId);
+    let entity = AirProfile.load(id);
+    if (entity == null) {
+      entity = new AirProfile(id);
+      entity.name = name;
+      entity.user = userId;
+      entity.tokenId = tokenId;
+      let airToken = getOrCreateAirToken(chainId, tokenAddress);
+      airToken.save();
+      entity.tokenAddress = airToken.id;
+      if (extraIds.length > 0) entity.extras = extraIds;
+      entity.createdAt = block.id;
+      entity.lastUpdatedAt = block.id;
+    }
+    return entity as AirProfile;
+  }
+
+  /**
+   * @dev this function does not save the returned entity
```

Comment:
This is a create function, it can save the entity.

Reply:
done